### PR TITLE
WIP - DO NOT MERGE - Orc merge

### PR DESCRIPTION
**Pull request by dougwig**

### CODE REVIEW

File: a10_neutron_lbaas/a10_config.py [hunk 1]
```diff
@@ -28,7 +28,31 @@
...
+            self._config = config
+            self._load_config()
+            return
+
+        self._config_dir = self._find_config_dir(config_dir)
+        self._config_path = os.path.join(self._config_dir, "config.py")
+
+        real_sys_path = sys.path
+        sys.path = [self._config_dir]
+        try:
```
Comment (Cedev) [inline a10_neutron_lbaas/a10_config.py]: We're going to need to switch to runpy for this to work with scaling. Nuking sys.path prevents the config.py from importing anything. Doesn't prevent merging this pr.

File: a10_neutron_lbaas/a10_openstack_lb.py [hunk 2]
```diff
@@ -37,31 +38,48 @@
...
     def __init__(self, openstack_driver,
-                 plumbing_hooks_class=hooks.PlumbingHooks,
+                 plumbing_hooks_class=None,
                  neutron_hooks_module=None,
-                 barbican_client=None):
+                 barbican_client=None,
+                 config=None,
+                 config_dir=None,
+                 caller_data=None,
+                 client_wrapper_class=None
```
Comment (Cedev) [inline a10_neutron_lbaas/a10_openstack_lb.py]: What's the purpose of `client_wrapper_class`?

File: a10_neutron_lbaas/db/migration/alembic_migrations/versions/2a280aba7701_create_orchestration.py [hunk 3]
```diff
@@ -56,16 +56,17 @@ def upgrade():
...
-    # )
+    op.create_table(
+        'a10_slbs',
+        sa.Column('id', sa.String(36), primary_key=True, nullable=False),
+        sa.Column('created_at', sa.DateTime),
+        sa.Column('updated_at', sa.DateTime),
+        sa.Column('tenant_id', sa.String(36), nullable=False),
+        sa.Column('device_name', sa.String(1024), nullable=False),
+        sa.Column('pool_id', sa.String(36)),
+        sa.Column('vip_id', sa.String(36)),
```
Comment (Cedev) [inline a10_neutron_lbaas/db/migration/alembic_migrations/versions/2a280aba7701_create_orchestration.py]: What's `vip_id` for?

File: a10_neutron_lbaas/plumbing_hooks.py [hunk 4]
```diff
@@ -111,19 +112,57 @@ def select_device(self, tenant_id):
 
 class VThunderPlumbingHooks(PlumbingHooks):
 
-    def select_device(self, tenant_id, db_session=None):
+    def select_device_with_lbaas_obj(self, tenant_id, a10_context, lbaas_obj,
```
Comment (Cedev) [inline a10_neutron_lbaas/plumbing_hooks.py]: do we need a10_context for anything?

File: a10_neutron_lbaas/plumbing_hooks.py [hunk 5]
```diff
@@ -111,19 +112,57 @@ def select_device(self, tenant_id):
...
+            root_id = lbaas_obj.root_loadbalancer.id
+            slb = models.A10SLB.find_by_loadbalancer_id(root_id, db_session=db_session)
+            if slb is not None:
+                d = self.driver.config.get(slb.device_name, db_session=db_session)
+                if d is None:
+                    raise ex.InstanceMissing(
+                        'A10 instance mapped to loadbalancer_id %s is not present in db; '
+                        'add it back to config or migrate loadbalancers' % root_id)
+                return d
+        else:
```
Comment (Cedev) [inline a10_neutron_lbaas/plumbing_hooks.py]: This is a little inflexible. What if professional services needs to do device per tenant for lbaasv2?

File: a10_neutron_lbaas/a10_config.py [hunk 6]
```diff
@@ -28,7 +28,31 @@
...
+            try:
+                sys.modules.pop('config', None)
+                import config
+                self._config = config
+            except ImportError:
```
Comment (Cedev) [inline a10_neutron_lbaas/a10_config.py]: `except ImportError` eats the import errors when trying to import `a10_neutron_lbaas.plumbing_hooks` to use the `VThunderPlumbingHooks`.

File: a10_neutron_lbaas/tests/unit/vthunder/test_instance_manager.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+import a10_neutron_lbaas.a10_exceptions as a10_ex
+import test_base
+
+# Figure out why this is necessary to make the mock work correctly.
+
+log_mock = mock.MagicMock()
+
+# This is here because I couldn't get mock to actually mock the classes
+# Get rid of this and I'll give you a cookie - MD
```
Comment (mdurrant-b3) [inline a10_neutron_lbaas/tests/unit/vthunder/test_instance_manager.py]: No one wanted a cookie?  In seriousness, this is awful and mock should be able to handle this with `mock.patch` at the class level.  It works so don't fix what ain't broken, but if there's a better way I'm all ears.

File: test-requirements.txt [hunk 8]
```diff
@@ -1,4 +1,11 @@
 fixtures>=1.3.1
 hacking>=0.10,<0.11
 mock>=1.0.1,<1.1.0
+mox
+mox3
 nose
+nose-exclude
+
+# TODO(dougwig) -- can we remove this?
+-e git+https://github.com/openstack/neutron#egg=neutron
```
Comment (mdurrant-b3) [inline test-requirements.txt]: If we mock the whole neutron module in tests, I believe so.  It appears the only reason it's not is we're using constants from neutron.

**Comment by Cedev** [comment]
Defaults for config aren't working correctly:

```
[1:57] 
 ```2016-05-18 13:55:41.134 TRACE neutron   File "/usr/local/src/a10-neutron-lbaas/a10_neutron_lbaas/a10_config.py", line 169, in get
2016-05-18 13:55:41.134 TRACE neutron     return getattr(self._config, key)
2016-05-18 13:55:41.134 TRACE neutron AttributeError: 'ConfigModule' object has no attribute 'verify_appliances
```

File: a10_neutron_lbaas/etc/defaults.py [hunk 9]
```diff
@@ -52,3 +57,21 @@
...
+VTHUNDER_REQUIRED_FIELDS = [
+    'username',
+    'password',
+
+    'nova_flavor',
+    'glance_image',
+    'glance_image_tag',
+
+    'vthunder_tenant_id',
+    'vthunder_tenant_username'
```
Comment (Cedev) [inline a10_neutron_lbaas/etc/defaults.py]: This is missing a `,` at the end of the line.

File: a10_neutron_lbaas/neutron_ext/extensions/a10Appliance.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+from neutron.common import exceptions
+import neutron.i18n as localization
+from neutron.services import service_base
+
+import a10_neutron_lbaas.neutron_ext.common.constants as constants
+from a10_neutron_lbaas.neutron_ext.common import resources
+import a10_neutron_lbaas_client.resources.a10_appliance as a10_appliance
+
+# Neutron is finicky. Sometimes _ is defined, sometimes it isn't
+localization.install()
```
Comment (mdurrant-b3) [inline a10_neutron_lbaas/neutron_ext/extensions/a10Appliance.py]: `_` can be imported from neutron.i18n as:
`from neutron.i18n import _`

**Comment by Cedev** [comment]
Trying to use the VThunderPlumbingHooks:

```
  File "/usr/local/src/a10-neutron-lbaas/a10_neutron_lbaas/plumbing_hooks.py", line 135, in select_device_with_lbaas_obj
    slb = models.A10SLB.find_by_loadbalancer_id(root_id, db_session=db_session)
AttributeError: type object 'A10SLB' has no attribute 'find_by_loadbalancer_id'
```

File: a10_neutron_lbaas/plumbing_hooks.py [hunk 11]
```diff
@@ -173,11 +173,13 @@ def select_device_with_lbaas_obj(self, tenant_id, a10_context, lbaas_obj,
...
         device_config = {}
         for key in vth:
+            if key in ['status', 'ha_sync_list']:
+                continue
             if key in defaults.DEVICE_REQUIRED_FIELDS or key in defaults.DEVICE_OPTIONAL_DEFAULTS:
                 device_config[key] = vth[key]
         device_config.update({
+            'tenant_id': tenant_id,
             'nova_instance_id': instance['nova_instance_id'],
-            'host': instance['ip_address'],
```
Comment (Cedev) [inline a10_neutron_lbaas/plumbing_hooks.py]: No idea why I deleted this ... added it back in next set of changes.

**Comment by dougwig** [comment]
Merged into origin/orc-merge

**Pull request closed without merging**